### ISSUE TITLE
strictNullChecks - compiler incorrectly detects "if" / "typeof" flow

### ISSUE THREAD

**Issue by Proculopsis** [Microsoft/TypeScript#20099]
<code>
/* 
Compiler error in this sample code on TypeScript Playground: 
https://goo.gl/zwJK6a 

Select "strictNullChecks" on the Options tab
*/

function strictNullChecks(delegate?: (message: string) => void) {
  let state = `typeof delegate = "${typeof delegate}"`;

  if (false == (typeof delegate === "function"))
    alert(state);
  else
    delegate(state); // typeof delegate is always "function"
}

strictNullChecks(window.alert);
strictNullChecks();

/*
There seems to be an issue with the "strictNullChecks" 
compiler option when passing an optional delegate 
parameter.

The above code works correctly but the compiler reports an
"Object is possibly 'undefined'." warning for parameter 
"delegate".

It seems to be checking for the reference being bracketed 
by an "if" statement checking for "typeof delegate" but
ignores that the equality might be compared to "false".
*/
</code>

Expected behavior: No compiler errors.

Actual behavior: Error reported in "else" clause of "if" statement.

**Comment by andy-ms** [comment]
Here's a simple repro:
```ts
function f(x: number | string): string {
    return (typeof x === "number") === false ? x : "";
}
```
We don't recognize the `(typeof x === "number") === false` -- we will only handle direct tests of `typeof x === "number"` or `typeof x !== "number"` (which are easier to read anyway).

**Comment by Proculopsis** [comment]
More playground sample code here: https://goo.gl/M2oWU4

Thanks Andy for the explanation - if you view the above code, it looks like the compiler is simply pattern matching for 'typeof' / '===' / '!==' rather than evaluating the 'if' clause, if on-the-fly evaluation is even possible, although 'if#5' is correctly interpreted which is logically the same as 'if#1' and 'if#2'.

**Comment by RyanCavanaugh** [comment]
We can't realistically backsolve every expression involving `typeof` to figure out what the types should be as a result. Comparing to booleans in `if` conditions doesn't occur often enough in real code to justify spending the performance budget to look for it.
